### PR TITLE
[fix]: [PL-28103]: Handle GCP KMS UnathenticatedException in Secrets …

### PR DESCRIPTION
**Pull request by Raghav-Murali**
…Management

## Describe your changes
This PR handles the Unauthenticated Exception thrown by the GCP KMS incase of invalid credentials. Previously this exception was not explicitly caught and thus a generic message was displayed in the UI which doesn't convey any valid information to the end user or providing any details that it was the credentials that were invalid. This change will catch the exception explicitly and propagate back the error message and code which conveys to the user that something was wrong with the provided credentials.

[Test Scenario] - Create a new GCP KMS secret manager connector with given credentials. Since the provided credentials are no longer valid it doesn't allow to create a connector.
[Test Outcome] - GCP KMS doesn't recognize the given credentials and thus it throws a “UnauthenticatedException”.

[Prior Error Message on UI]-
Encryption failed after 3 retries

[New Error Message on UI with this fix]-
Unauthenticated exception thrown while encrypting secret. Error code - GrpcStatusCode{transportCode=UNAUTHENTICATED} Error Msg - io.grpc.StatusRuntimeException: UNAUTHENTICATED: Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.

As seen above the error code and message contained in the UnauthenticatedException provides sufficient details on what needs to be corrected.

## Checklist
- [Yes ] I've documented the changes in the PR description - Check [PL-28103](https://harness.atlassian.net/browse/PL-28103) for sufficient summary.
- [ Yes ] I've tested this change either in PR or local environment - Tested in the local test environment.
- [ Yes ] I've added unit tests to verify the changes - Added sufficient unit test that covers all possible scenarios for this change.
- [ N/A ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/37371)

### CODE REVIEW

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_Failed_** 
 


            
 #### 2 Issues 
* Code issue 2 Code issues 

 [Details](https://app.embold.io/organization/gh/harness/repositories/7752dbb3099d2dba3b5d37f5931a5e4c/pull-request/37371)

**Review by Raghav-Murali — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37371)* status: 3 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`890-sm-delegate/src/main/java/io/harness/encryptors/clients/GcpKmsEncryptor.java` line 104 at r1](https://reviewable.io/reviews/harness/harness-core/37371#-NBiT03u6hW-quWXeKJI-r1-104:-NBrnuQT9un352NLmq2y:bw6ngop) ([raw file](https://github.com/harness/harness-core/blob/cd07dcafd6832e45918128a9f3482e9ae5c5c36b/890-sm-delegate/src/main/java/io/harness/encryptors/clients/GcpKmsEncryptor.java#L104)):*
<details><summary><i>Previously, rickmarry (Rick Marry) wrote…</i></summary><blockquote>

Do we really want to give all this information in the response in the UI?  I think that stating the we have failed to authenticate with  GCP KMS secret manager should be enough
</blockquote></details>

I will remove the additional message in the front and just send the message in the exception which gives all the details.

File: 890-sm-delegate/src/main/java/io/harness/encryptors/clients/GcpKmsEncryptor.java [hunk 1]
```diff
@@ -95,6 +96,15 @@ public EncryptedRecord encryptSecret(String accountId, String value, EncryptionC
       try {
         return HTimeLimiter.callInterruptible21(timeLimiter, Duration.ofSeconds(DEFAULT_GCP_KMS_TIMEOUT),
             () -> encryptInternal(accountId, value, gcpKmsConfig));
+      } catch (UnauthenticatedException e) {
+        log.warn("Unauthenticated exception thrown while encrypting secret");
```
[the remark this answers is not in the document]
Comment (meenakshi-harness) [reply 890-sm-delegate/src/main/java/io/harness/encryptors/clients/GcpKmsEncryptor.java:100]: Also should we add this log only when it is not retry-able?

File: 890-sm-delegate/src/main/java/io/harness/encryptors/clients/GcpKmsEncryptor.java [hunk 2]
```diff
@@ -166,6 +176,15 @@ public char[] fetchSecretValue(String accountId, EncryptedRecord encryptedData,
           return HTimeLimiter.callInterruptible21(timeLimiter, Duration.ofSeconds(DEFAULT_GCP_KMS_TIMEOUT),
               () -> decryptInternal(encryptedData, gcpKmsConfig));
         }
+      } catch (UnauthenticatedException e) {
+        log.warn("Unauthenticated exception thrown while decrypting secret");
```
Comment (meenakshi-harness) [inline 890-sm-delegate/src/main/java/io/harness/encryptors/clients/GcpKmsEncryptor.java:180]: same

File: 890-sm-delegate/src/main/java/io/harness/encryptors/clients/GcpKmsEncryptor.java [hunk 3]
```diff
@@ -95,6 +96,15 @@ public EncryptedRecord encryptSecret(String accountId, String value, EncryptionC
       try {
         return HTimeLimiter.callInterruptible21(timeLimiter, Duration.ofSeconds(DEFAULT_GCP_KMS_TIMEOUT),
             () -> encryptInternal(accountId, value, gcpKmsConfig));
+      } catch (UnauthenticatedException e) {
+        log.warn("Unauthenticated exception thrown while encrypting secret");
+        if (!e.isRetryable()) {
+          String errorMsg = e.getMessage();
+          String errorCode = e.getStatusCode().toString();
+          String message = "Unauthenticated exception thrown while encrypting secret. Error code - " + errorCode
```
[the remark this answers is not in the document]
Comment (meenakshi-harness) [reply 890-sm-delegate/src/main/java/io/harness/encryptors/clients/GcpKmsEncryptor.java:104]: +1
For UI we can format same information in HInt and  Explanation format if possible like this
`hintWithExplanationException(HintException.HINT_GCP_ACCESS_DENIED,
                ExplanationException.INVALID_PARAMETER,
                new InvalidRequestException(e.getMessage(), GCP_KMS_OPERATION_ERROR, USER))`

**Review by meenakshi-harness — CHANGES_REQUESTED** [review]

File: 890-sm-delegate/src/main/java/io/harness/encryptors/clients/GcpKmsEncryptor.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Raghav-Murali) [reply 890-sm-delegate/src/main/java/io/harness/encryptors/clients/GcpKmsEncryptor.java:104]: Cool, makes sense. Will update it accordingly. Thanks guys!

File: 890-sm-delegate/src/main/java/io/harness/encryptors/clients/GcpKmsEncryptor.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Raghav-Murali) [reply 890-sm-delegate/src/main/java/io/harness/encryptors/clients/GcpKmsEncryptor.java:100]: Yeah, I can change it to an error and log only when its non-retry able.

File: 890-sm-delegate/src/main/java/io/harness/encryptors/clients/GcpKmsEncryptor.java (hunk 2 above)
Comment (Raghav-Murali) [reply 890-sm-delegate/src/main/java/io/harness/encryptors/clients/GcpKmsEncryptor.java:180]: Done

**Review by Raghav-Murali — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37371)* status: 3 file(s) to review. 3 change requests. 7. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`890-sm-core/src/test/java/io/harness/encryptors/GcpKmsEncryptorTest.java` line 175 at r2](https://reviewable.io/reviews/harness/harness-core/37371#-NC1psrs0Ca0tkZPCX8_:-NC1zWfwD6ya16yQTvMd:b7m9wun) ([raw file](https://github.com/harness/harness-core/blob/2064a400f6628079fb6451276dce5db66621ed70/890-sm-core/src/test/java/io/harness/encryptors/GcpKmsEncryptorTest.java#L175)):*
<details><summary><i>Previously, gn-harness (Gaurav Nanda) wrote…</i></summary><blockquote>

do we need test for retriable UnauthenticationException too?

If this can never be retriable -- isRetriable() check is of any value?
</blockquote></details>

Yes, I will add another unit test where isRetriable() is true. 

Similar to in Azure Client the GCP client also has internal retry mechanism. And I think based on that it will either set this to true or false.
___
*[`890-sm-delegate/src/main/java/io/harness/encryptors/clients/GcpKmsEncryptor.java` line 100 at r2](https://reviewable.io/reviews/harness/harness-core/37371#-NC1oeib16Q0IT7tFUOa:-NC2-VOs6SX8cdFJLeoV:b-awcbnd) ([raw file](https://github.com/harness/harness-core/blob/2064a400f6628079fb6451276dce5db66621ed70/890-sm-delegate/src/main/java/io/harness/encryptors/clients/GcpKmsEncryptor.java#L100)):*
<details><summary><i>Previously, gn-harness (Gaurav Nanda) wrote…</i></summary><blockquote>

What to do when this is retriable?

Is it possible that we retry this three times and then come back in UnauthenticatedException block again? 
</blockquote></details>

I previously thought if its retryable then to keep trying again and again. As per the docs - https://cloud.google.com/java/docs/reference/google-cloud-core/latest/com.google.cloud.BaseServiceException#com_google_cloud_BaseServiceException_isRetryable__ this function is suppose to return true only if its safe to retry. And in other cases it should return false. But the docs doesn't mention if there is a limit on how many time it will return True.

Yes, now there is no three times limit for this UnauthenticatedException block, so it will retry as many times as isRetryable() returns true. I'm assuming the GCP client will not indefinitely just keep returning true. But a safer approach is to do the three times approach. Let me modify it. 
___
*[`890-sm-delegate/src/main/java/io/harness/encryptors/clients/GcpKmsEncryptor.java` line 101 at r2](https://reviewable.io/reviews/harness/harness-core/37371#-NC1p_fu4-XbD-Uh-hW0:-NC21H1s-aMrSVOGv2hZ:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/2064a400f6628079fb6451276dce5db66621ed70/890-sm-delegate/src/main/java/io/harness/encryptors/clients/GcpKmsEncryptor.java#L101)):*
<details><summary><i>Previously, gn-harness (Gaurav Nanda) wrote…</i></summary><blockquote>

Please log the exception too
</blockquote></details>

Done.
___
*[`890-sm-delegate/src/main/java/io/harness/encryptors/clients/GcpKmsEncryptor.java` line 180 at r2](https://reviewable.io/reviews/harness/harness-core/37371#-NC1phmA2fl3wOKUJ6V1:-NC21Fyt9gqITEgTOuUX:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/2064a400f6628079fb6451276dce5db66621ed70/890-sm-delegate/src/main/java/io/harness/encryptors/clients/GcpKmsEncryptor.java#L180)):*
<details><summary><i>Previously, gn-harness (Gaurav Nanda) wrote…</i></summary><blockquote>

same comments as above
</blockquote></details>

Done.

**Review by meenakshi-harness — DISMISSED** [review]
> Reviewed 1 of 2 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37371)* status: Some of the checks are too old. 3 file(s) to review. 2 change requests. 5. unresolved discussions. 1 of 2 approvals obtained.

**Comment by Raghav-Murali** [comment]
trigger feature-build

**Review by gn-harness — DISMISSED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37371)* status: 2 file(s) to review. 1 change requests. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by rickmarry — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37371#-:-NCXWkfhAk7kymJS50mZ:bnfp4nl)* status: 2 file(s) to review. 1. unresolved discussions.

**Review by rickmarry — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37371)* status: 1 change requests. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by rickmarry — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37371#-:-NCXWuvH18rSkydJGpqY:bnfp4nl)* status: 1. unresolved discussions.

**Comment by Raghav-Murali** [comment]
trigger feature-build

**Review by meenakshi-harness — APPROVED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37371)* status: 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by gn-harness — APPROVED** [review]
> Reviewed 2 of 2 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37371)* status: 1. unresolved discussions.

File: 890-sm-delegate/src/main/java/io/harness/encryptors/clients/GcpKmsEncryptor.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (rickmarry) [reply 890-sm-delegate/src/main/java/io/harness/encryptors/clients/GcpKmsEncryptor.java]: good.. thanks

**Review by rickmarry — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37371)* status: :shipit: complete! ✓

**Pull request merged**